### ISSUE TITLE
fingerprinting

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#620]
I want to address this [reddit post](https://old.reddit.com/r/firefox/comments/ajckk1/is_resist_finger_print_making_me_more_unique/)

First of all, test sites like Panopticlick etc are flawed, especially when returning entropy. I'm not knocking the sites or what they do to raise awareness etc, just providing some context that almost everyone I see commenting on reddit and elsewhere get completely wrong. The information returned is all you should be concerned about - i.e: see what is returned by the browser. Generally speaking, the data sets are
- not real-world (that's not the test site's fault, that's the nature of providing such a site)
   * the tests attract those interested in the topic, and are instantly biased/compromised
       * e.g Firefox in many of these tests is up as high as 35-40%
   * users trying to lower their entropy score (tiny changes, repeat tests)
       * this pollutes the data set
   * repeated tests of the above users
       * this lowers the entropy of some metrics
   * it is a [self fulfilling prophecy](https://en.wikipedia.org/wiki/Self-fulfilling_prophecy): let's take for example that canvas (unaltered) is always a unique value. This instantly makes every user unique, and it will always be the highest entropy item. This makes users research, ask questions and leads them to blocking their canvas, or using RFP (or random spoofing) - with repeated tests. Now we have the situation where RFP canvas is (from memory) about 1 in 30. This is INSANE. I call BULLSHIT that 3.3% of the world uses RFP. Are you seriously saying that of the 6% of users worldwide who use Firefox/Tor Browser... that over half of them have RFP enabled? Come on! (note: this is not a *perfect* example, because other browsers and extensions can also return the same 10x10 white canvas that RFP does, but I bet it's insignificant)
- not current
   * they incorporate old data e.g. real world metrics on UA suffers
   * some sites allow a time range, others (I think Panopticlick did this) reset their data set
- not relevant
   * they could use real world metrics such as 6% are Firefox, and extrapolate from that
- small (especially if you were to filter results for a recent time frame)
- limited (some tests need upgrading, others need to be introduced)

I was going to provide some examples but cannot get Panopticlick to finish loading the results. I did a few days ago, but it's playing up again/

I'm too busy to find the link, but there was (details from memory) a real world study done approx a year ago  in France, using millions of visitors on a popular site(s). **This** is how you get metrics. The results showed up a lot of flaws in people's thinking IMO.

Use math and science, as the Tor Uplift and Tor Project do, and you'll find the real world metrics fall into line. The real world test above validated this IMO. I should find the paper for you.

---

This [comment](https://old.reddit.com/r/firefox/comments/ajckk1/is_resist_finger_print_making_me_more_unique/eevsgpj/) needs addressing
> Absolutely useless. The whole user.js thing makes you more unique and as thus easier identifiable (imagine how many people use that specific user.js over people who "just" have a Tor fingerprint).
> 
> You achieve privacy by looking like everybody else not by changing as much as possible. The more you change the more unique you actually become. That is why I am saying just toggle RFP, Container tabs and disable webgl and nothing else. That is also why Tor says to not change the defaults.
> 
> Afaik RFP should make FF identical to Tor for websites (though not sure need a dev verification here). If you trust Tor then it is enough. If you don't trust Tor then there is literally no escape for you.

Let me start by saying that FP'ing is overrated - but becoming more mainstream since the rise of adblockers in the last few years - e.g reddit now includes a first party version of fpjs2. You control FP'ing by controlling what you allow to run in your browser. 95% of the FP attack surface is third parties and a couple of FP items (e.g canvas - because it's so easy to drop in, and it makes you unique). The tracking industry are the ones who drive this in the wild, and they have so much lower hanging fruit, that it is not worth their time and effort to use many metrics (not that we shouldn't keep them in mind and build defenses against them).

Also, looking at things such as mixed content (which can be controlled on a site by site basis) are pretty much useless metrics from a tracking perspective. So are x,y screen / window / inner window position metrics, so are cookie & persistent storage permission metrics, so are permission metrics (geo, notifications etc - prompt/block) and I could name more. They just aren't consistent enough.

> Absolutely useless [subsequent post: Though keep in mind I am talking about that specific user.js.]

To the author of that comment: I sincerely hope you didn't mean the user.js, because if you did, then I think you have completely misunderstood the purpose of this user.js. If you were replying to the strategy of using RFP and falling back to Canvas Blocker in fake mode, then I fail to see how that is useless. It's very valid. If you are using RFP then you need to look like the set of the RFP users, and this strategy allows you to use sites that *need* canvas to work but still not uniquely FP you.

> The whole user.js thing makes you more unique and as thus easier identifiable

Every single browser (except Tor Browser, I'll address that later) is unique when you do **nothing**. The user.js is can't do everything. It's limited in scope. First and foremost you want anonymity, privacy and security (don't confuse what these are). That means a trade-off. By blocking for example SSL Session IDs - you are already in a super tiny uber small group of users, should that be used as a metric. But allowing it can totally betray you (e.g, you search google, you change VPNs, you search google again). So it's a question of : do you value anonymity over the *possible* tracking of you by linking various web activities? SSL Session IDs is a server side tracking rather than a FP, but it still a metric, so maybe not the best example of a FP metric. But the main point stands - that anonymity, privacy and security come first. FP'ing doesn't necessarily damage these either, but is of course addressed where possible. It's always about trade-offs.

Control what happens client side. For myself (uMatrix) I only allow first party css and images - I think gorhill calls it nightmare mode. Even so, just using uBO with default filters kills a lot of this stuff.

We explicitly **tell users to use extensions** in the user.js and provide a wiki and suggest extensions, and rules, and so on.

> The more you change the more unique you actually become ...

Do almost nothing to your browser except turn on RFP. You are still unique (or extremely close to it). I can detect your real FF version via feature detection, I can get your font fingerprint including if you disable sites using their own fonts, and disable JS (I can grab them via CSS, and a *lot* of them). I could go on.

Everyone (Firefox) is already unique (or very close to it) regardless of what you do - IF sites and trackers wanted to get that - hence .. reduce the attack surface is your best defense, and have **some** fallbacks - the user.js here actually **doesn't** disable or change a lot of things for this reason eg we don't disable gamepads because it is a metric, we don't disable geo (which is behind a prompt) because that's another bit of entropy, and so on.

> That is why I am saying just toggle RFP, Container tabs and disable webgl and nothing else

RFP is not enough in Firefox. It is no where near being complete. You are still unique. Containers do nothing for FP'ing. WebGL helps .. but OMG .. you are now in a super tiny percentage worldwide: FF users with no WebGL .. now add RFP to that and you are even more unique .. now let me get some more metrics you didn't think of .. you ARE unique (still)

You will not defeat FP'ing (if it really wanted to) in Firefox (yet). Control what happens client side. Reduce the attack surface.

> You achieve privacy by looking like everybody else not by changing as much as possible

You have to change **shitloads** in order to look like everybody else .. no wait a minute .. **everybody else is unique** - there is no "everyone else". What you are trying to do is **lower entropy** and that only works when you enforce a set of users - such as Tor Browser and Firefox RFP users (in the user.js we actually MAKE SURE that people don't change prefs that affect the RFP fingerprint - see section 4600). RFP is no-where near finished.

Also, this really has nothing to do with privacy. The word you are looking for is anonymity, although it's more tracking than anything else (you may have never leaked you real ID in this particular setup). Please don't mix these up. Anonymity means never linking back to the real you. Privacy means no-one else but the sending and receiving party can read it. And tracking is all about linking activities together (which can break your anonymity).

> over people who "just" have a Tor fingerprint

**Only Tor Browser (TB) users have a Tor fingerprint**, and there are a few of them because not everything can be spoofed. You cannot mask this. It is easily detectable no matter what you do. TB is not trying to hide the fact that it is TB, and FF is not trying to hide the fact that it is FF, even with RFP - and no matter what you do or change to look like TB, you will always be FF, and easily spotted. So now you are in a super extraordinarily tiny minuscule group if not actually unique (again, since you were n a default FF even with RFP on). 

Again, TB is a set where all the rules can be enforced, so the entire set of users look the same (as much as possible, e.g you can still easily detect the platform with a simple 3 font check - so there will always be some fragmentation). RFP on Firefox is not "enforced" (yet), e.g some prefs will alter the fingerprint.

The user.js actually says, at the top (and in our short readme here at GitHub), to consider using TB if it meets your needs. This user.js is **not trying to be the Tor Browser** sans Tor protocol.

> Afaik RFP should make FF identical to Tor for websites

Not even close - years to go. And that is not the purpose of Tor Uplift. TB can enforce things, FF can't. TB can disable APIs, FF can't because it is a mainstream browser that needs to work with standards. For this reason, RFP can only do so much (for now) - sure, they could make RFP disable WebGL, disable Audio API, and so on - but then there would be no uptake by FF users when they give it a UI presence - and that damages the numbers - you need a sizeable set of users with the same FP in those metrics covered by RFP, in order for RFP to be effective.

I think I've typed enough for today (understatement of the week)

**Comment by KOLANICH** [comment]
>Privacy means no-one else but the sending and receiving party can read it.

about it.

>Only Tor Browser (TB) users have a Tor fingerprint

There is ain't no thing as a Tor Browser fingerprint.

>is not trying to hide the fact that it is TB, and FF is not trying to hide the fact that it is FF, even with RFP - and no matter what you do or change to look like TB, you will always be FF, and easily spotted.

Isn't TB a Firefox? So if all the patches uplifted, ff will be differrent from TB only by prefs and logo. Changing the prefs will effectively make TB and FF the same.

**Comment by Thorin-Oakenpants** [comment]
> Privacy means no-one else but the sending and receiving party can read **about** it.

No. That is not true. I can *observe* the metadata, just not the contents.

> There is ain't no thing as a Tor Browser fingerprint.

Well there is, since I was directly referring to the reddit comment. And I did point out that even in TB, there are variations in results - thus fragmentation. But for all intents and purposes - there is an overall  TB FP for lack of a better word. It's just that if you dig deep enough, you can find differences (platform for example) .. but that, AFAIK, for now, that still doesn't make you unique, and at worse, I suspect one of about 6 or 9 TB FP buckets (it would need a bit more investigation and testing)

> Isn't TB a Firefox

It's based on ESR, so it rapidly falls behind stable. The core RFP still applies, but it's up to TB what they do, when they get around to it - e.g they still apply some of the prefs in our 4600 section. Plus they have a shit-load of their own patches, not to mention a fuck-tonne of pref fiddling. But yeah, the idea is that RFP reduces the amount of rebasing, but no - they're still wildly divergent

**Comment by crssi** [comment]
I am thinking now for a long time already that FPing is quite useless, if there is no tracking.
I don't say that FP should not be addresses (quite opposite), but if for example EvilCorp is unable to track you (and your browsing), then FP has lower value (almost none).

@Thorin-Oakenpants, Am I wrong?

**Comment by crssi** [comment]
Do you see WebGL as security risk or FP risk or both?

**Comment by Thorin-Oakenpants** [comment]
> I am thinking now for a long time already that FPing is quite useless, if there is no tracking.

Wot? That's what FP actually is... the ability to track a specific browser when all other tracking methods are nullified (e.g cookies / persistent storage deleted, logins not used, first party isolation, using Tor protocol, no cacahes, no etags, blah blah)

**Comment by crssi** [comment]
> Wot? That's what FP actually is... the ability to track a specific browser when all other tracking methods are nullified

I understand that, but when FP values (also referers, origins, etc.) are not passed to 3rd party then there the tracking is very limited or none.
So in reverse order... if tracking/sharing is eliminated, then FP has lower value.
I haven't forgot also that XHRs can be evil... but this is "extended" story.

**Comment by Thorin-Oakenpants** [comment]
> if tracking/sharing is eliminated, then FP has lower value

Rubbish! If they can't link your activities together by conventional means, then FP'ing becomes extremely valuable

**Comment by Atavic** [comment]
> a real world study done approx a year ago in France

Maybe [Beauty and the Beast](https://hal.inria.fr/hal-01285470v2/document)? They used data from AmIunique.org

...same data used by [P. Laperdrix](https://tel.archives-ouvertes.fr/tel-01729126/document) to
design two countermeasures called Blink and FPRandom.

**Comment by Atavic** [comment]
^^FP-Scanner: [The Privacy Implications of Browser
Fingerprint Inconsistencies](https://www.usenix.org/conference/usenixsecurity18/presentation/vastel)

*From reddit comment*:

> You achieve privacy by looking like everybody else

No, you don't. You're just into a big flock, moved around by Google and Cloudflare wolves.

> Do you see WebGL as security risk or FP risk or both?

Greatly a FP risk, it exposes your graphics drivers version, hence your GPU.

Re: WebGL Security, see: https://www.khronos.org/webgl/security/

**Comment by seanob86** [comment]
Not sure if here's the right place to query.. and, I'm not so technical on this FP discussion but is something I am aware of.

I have decided to set RFP to false. I've been having issues with this set to true and will re-visit it at a later time.

With RFP = false I have enabled section 4600 of the user.js, but I have also installed the "CanvasBlocker" extension with block mode configured to "fake readout API". The effect this has is that it randomizes the canvas fingerprint on each page that is visited/refreshed.

I have also read various posts on reddit which only seem to suggest two options (1) Enable RFP or (2) install an extension such as "CanvasBlocker" to randomize the Canvas Fingerprint.

So, with RFP disabled and have enabled the settings listed under 4600 of ghacks - user.js. Should I also be bothered with installing an extension such as "CanvasBlocker" to randomize the Canvas fingerprint?

Thanks in advance!

**Comment by Thorin-Oakenpants** [comment]
@crssi OK, I think I have deciphered your comment

> I understand that, but when FP values [edit: remove stuff that has nothing to do with FPing] are not passed to 3rd party then there the tracking is very limited or none. So in reverse order... if tracking/sharing is eliminated, then FP has lower value.

FP'ing is not tracking: it is simply the art of getting your browser's FP. That FP can be traded and sold and can link web activity - it doesn't matter who does it (1st or 3rd party), it only matters that it is done. If you reduce the attack surface and control what happens in your browser, you control most of this. I think this is what you, crssi, are trying to say - and I already said it, several times :)

from me
> Wot? That's what FP actually is... the ability to track a specific browser when all other tracking methods are nullified

also from me
> Rubbish! If they can't link your activities together by conventional means, then FP'ing becomes extremely valuable

It is an assumption, and clearly implied (because otherwise the sentences make no sense), that the FP data is collected. If you blocked it, then bravo! That's the whole point of anti-FP'ing.

**Comment by KOLANICH** [comment]
> you'll always return a unique value, but it will change every time it is asked = not identifiable.

IMHO the wrong approach.

1. additional fingerprinting vector, marking the use of the addon randomizi g the fingerprint.
2. may open window on exploiting PRGs for identification, if an insecure PRG is used.

IMHO a better approach is to generate a random identity for every tracking party and return it deterministically.

**Comment by Thorin-Oakenpants** [comment]
> IMHO the wrong approach.

I was referring to @seanob86 's situation, where as a non-RFP user, using CB, random faking is just fine (it is the default).

As an overall strategy, I do not agree with **raising entropy** at all - all that does is create more work. Look at all the situations you're trying to work through - what a mess! You should read the TB design doc - you're just making life hard for yourself and over-complicating everything

The best solution is **lowering entropy** - again, math and science (and logic!). RFP and a 10x10 white canvas. If anyone is serious about privacy et al, then any shortcomings from RFP just need to be dealt with. Seriously, it hardly breaks anything. If you really have to, use a secondary browser for a problem site.

As for detecting you are faking, Panopticlick already just added a simple two canvas compare.

**Comment by Thorin-Oakenpants** [comment]
Panopticlick is back up and running. So the RFP canvas is actually being reported as 1 in 7.5 ... which is completely ludicrous as a real-world figure?

Or lets look at WebGL .. 1 in 6.5 have it disabled. Or lets check out that half the world has en-US for language (pretty sure all the Chinese and non-English speakers of the world make that quite a bit inflated). Or that 1 in 12 browsers only have 2 wingding fonts from their list (in other words, 1 in 12 users block document fonts). Seriously .. 15% of users killed WebGL? .. 95% of users wouldn't even know what the F that was. And over 8% of people decided to block all fonts and make their web experience ugly? Come on. These numbers are so out of whack, you can't use them

One more "Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0" = 1 in 14.7 browsers have this. Considering 90% of FF users are on stable, not ESR, that's a stretch. Anyway, 1 in 14.7 browsers = 6.8% market share for ESR .. that's more worldwide share than FF actually has, or at best, about the same. But FF's own metrics say only about 10% are on ESR .. and some more % points for TB .. and it's still massively overstated.

Anyway .. end of rant .. hope you all get my point. Was hoping someone would point the dude on reddit here. Not the one I replied to here, but the original reddit OP. The one I replied to, I read his comments on reddit from the last 3 weeks plus - the guy has almost no clue what he's talking about, and is super rude and confrontational to say the least.

**Comment by KOLANICH** [comment]
>As an overall strategy, I do not agree with raising entropy at all - all that does is create more work.

Do you mean that spoiling the stuff with fake data, if it is cheap to cause its production, if the data is indistinguishable from the real one within the budget affordable to tracking parties should not be used?

IMHO quite contrary. Let's define privacy that's capability to hide the data and metadata the subject prefers to hide. So even if we can produce deterministic fingerprints, randomizing fingerprints, if they are unlinkable

1. causes tracking parties to waste space collecting fake data and computation resources on analysing it. But should be done with care  - simulators can be fingerprinted themselves.
2. spoils machine learning models performance and metrics.
3. randomizing your results would result for less accurate predictions for user's case


1 + 2  makes the whole industry more costly to operate.
2 + 3 makes it's harder to extract useful information, so better privacy for the end user
1 + 2 + 3 makes the whole industry less profitable, so less investment into tracking peop

Though there is a flaw - it may be hard to cheaply and securely implement randomization because defeating it is primary area of tracking business, so they would have the researchers and resources to train the models, unlike Mozilla (though they have done some researches on ML, I have not seen such a progress on anti-tracking features), but even eliminating stockpiling of tracking data by everyone smaller than government-sponsored orgs should be beneficial.

**Comment by Thorin-Oakenpants** [comment]
^^ You're not looking at the problem, you're just creating extra one(s). Assume the worst case scenario - that tracking companies will do whatever they have to in order to get more data, money is no object, they have a gazillion jillion bazillion trillion dollars. In other words, **just look at the original problem**.

Problem: canvas = unique
Solution1: lower entropy: in sufficient numbers (eg in the subset of TB or RFP users)
Solution2: raise entropy: always randomize

Both of these are just as effective. Both are easily detected (if they wanted to) as some entropy. i.e is the canvas a white 10x10 (yes/no) or is the canvas randomized on each request (yes/no). But far better than being unique.

Done: the problem has been solved

---

**You're trying to solve a different problem**, one of your own making, that is blown out of proprtion IMO - hiding the fact that you randomly fake. Subsets such as TB or RFP users don't care, because they are all the same. So this is only a problem for those who raise entropy. **Forget about the tracking companies, focus on the problem**

Problem: I give away a tiny bit of entropy that I always fake a random canvas
Solution: re-use the randomized canvas value: This brings a shitload of complexity and edge cases etc
  - per first party domain? but what if load youtube as 3rd party on two sites?
  - per domain? but what if close all those tabs, but then revisit?
  - per tab? but if I open two tabs on reddit, then reddit can still tell I fake
  - per session? but what if I change VPNs mid session?
  - per container?
  - what happens in PB mode?
  - what happens with session restore?
  - maybe use a timer? no, this is just stupid
  - what about sites I log into? If I randomize each session they will still know. Wait, I'm logged in. Phew. But now they know (and could share that). Better build in something else to always return the same fake canvas per site but only when logged in. I better store than info somewhere. Also, OMG .. how do I tell if you're logged in? I also better return this same value in third party requests.

And it goes on? I'm barely scratching the surface. There are so many ways your canvas can be shared.

This is the problem with raising entropy. You then need to tell lies, and then more lies, and then eventually you slip up. Seriously, just go read the TB design doc about lowering vs raising entropy

**Comment by KOLANICH** [comment]
>per first party domain? but what if load youtube as 3rd party on two sites?

Yes, per first-party webapp. Webapp is determined by cookies sharing and CORS. Each webapp gets a separate identity cleared "when cookies are cleared". Then each instance of youtube gets a different fingerprint. Each fingerprint is unlinkable to each other, because of different cookies sets. Of course there is a problem with IP address, so Tor should be used.

>but what if close all those tabs, but then revisit?

If one keeps cookies between sessions, he is tracked.

>but what if I change VPNs mid session?

new VPN will be linked to your new identity. It still will be since cookies are kept.

>what happens in PB mode?

Everything discussed was for PB mode. For non-PB mode there are cookies.

>what about sites I log into? If I randomize each session they will still know. Wait, I'm logged in. Phew. But now they know (and could share that). Better build in something else to always return the same fake canvas per site but only when logged in. I better store than info somewhere. Also, OMG .. how do I tell if you're logged in? I also better return this same value in third party requests.

1. We can introduce pinning.
2. There were addons like multifox, keeping separate cookie sets for each website, and they have been detecting signing in somehow (I have not analysed their source, but I guess they have detected signin forms), so they could detect and display usernames. So we can associate usernames with identities.

Even if pinning have failed, from the website PoV it would look like if a user have changed their browser. Not within a session, but each signin. (BTW, I constantly get messages from the services that I have changed my browser, it may mean thar RFP already incorporates some randomization).

**Comment by Thorin-Oakenpants** [comment]
sorry for sounding a bit snappy (got a lot of stuff going down, work wise)

The other things is, you would have to beat tampering detection: e.g against known pixel values - see  https://canvasblocker.kkapsner.de/test/ - take the Detection Test. I am not an expert, so I am not sure if this is preventable. I think kkapsner's CB methods prevents most of this (the canvas ones), but I would have to fire up a nilla FF with CB to test it. RFP does not.

**Comment by Thorin-Oakenpants** [comment]
^^

allow canvas and no CB (default; control group)
```
known pixel value test 1: no API tampering detected
known pixel value test 10: no API tampering detected
```


CanvasBlocker only
```
known pixel value test 1: no API tampering detected
known pixel value test 10: API tampering detected
```

RFP only
```
known pixel value test 1: API tampering detected
known pixel value test 10: API tampering detected
```

So looks like `known pixel value test 10` would break your super complex super-duper new-fangled whizzy do-dacky canvas fake-it-til-you-make-it i'm a real boy Pinocchio mechanism - even if you make it perfect

**Comment by Thorin-Oakenpants** [comment]
>  I should find the paper for you.

https://hal.inria.fr/hal-01718234/document

> We analyze 2,067,942 browser fingerprints collected from one of the top 15 French websites. The observations made on this novel dataset shed a new light on the ever-growing browser fingerprinting
domain

That's the paper, I'm sure. I'm going to re-read it

**Comment by Thorin-Oakenpants** [comment]
^^ I'll have yet to re-read it (e.g evaluate what FP metrics they looked at), but... right at the start [emphasis mine]

> The key insight is that the percentage of unique fingerprints in this dataset is much lower than what was reported in the past: **only 33.6% of fingerprints are unique** by opposition to over 80% in previous studies. We show that **non-unique fingerprints tend to be fragile. If some features of the fingerprint change, it is very probable that the fingerprint will become unique**

IMO, lowering entropy is the ONLY way to go. Raising entropy thru frequent randomization is a loser's game :trollface: trollface specially for @KOLANICH :)

I'll also point at this: https://hal.inria.fr/hal-01717158/document <- which is a little old, and they used RAS as an example (which is pre Quantum), but it shows that inconsistencies and the lies that are told by the extension were easy to spot. I can do the same to a lot of chameleon's spoofing, such as screen and timezone. And all that does is make you more unique.

Once again, all the experts (that I can find) have been saying for the five years or so I've been reading them ... to lower entropy. It's not quite that simple, but I'm not here to write a book (I think)

**Comment by KOLANICH** [comment]
>If some features of the fingerprint change, it is very probable that the fingerprint will become unique

IMHO not a problem, since the fake fingerprint is useless for linking profiles.

>but it shows that inconsistencies and the lies that are told by the extension were easy to spot.

Is there any papers where the lies were generated by a neural net?

**Comment by Thorin-Oakenpants** [comment]
> Is there any papers where the lies were generated by a neural net?

I don't know what you mean by that. Can you elaborate?

> IMHO not a problem, since a fake fingerprint is useless for linking profiles.

Did you read the paper? Of course, each item would need to be looked at in isolation, so I don't want to speak in generalizations. What I'm getting at is two fold

First: When you lie and get caught, that in itself is more entropy. That's OK if everyone in the group lies the same way - e.g TB users fibbing that their browser and screen are the same as the inner window. But take UA (JS navigator and HTTP headers), and lets say you lie about the browser (in your randomness) and claim to be chrome or safari - the lie is easily detected and you have more entropy. You will be one of a very very very very very very small group of FF users who randomize their UA, same with canvas (as you rightly point out, random faking on each request is a dead giveaway: whereas RFP is up front and everyone always looks the same). And I could go on, but each metric needs evaluation within it's own right.

Second: that when lying, it starts to get complicated. RFP patches get complicated too, when new features or APIs get added, e.g Intl.DateFormat. They're still trying to hide the real OS as much as possible. It's in the tor design doc somewhere. Patches that lower entropy but are plausible are so much  less work to maintain.

That said, there is nothing wrong in theory with randomizing constantly. It;s the same as lowering entropy - it needs large numbers ... AND ... you **still need to mask** the real values from other vectors.

What the paper says is that if someone changes a metric, they are more likely to become unique - i.e there are a small subset of various results that encompass most users: you know what I mean - e.g 10% of the fingerprints catered for 90% of the users. Fiddling with those in anyway would most likely make you unique - e.g see the example (from memory) of the accept header : change Windows8.1 FF to WIndows8.1 Chrome, but then you need to also change the http header.

They didn't say anything about randomizing it - I did, and that was probably the wrong way to put it.

----

Hey @KOLANICH - you got any androids lying around? Or MacOS? Or MacOS X?

If you do, can you go to my [test page](https://thorin-oakenpants.github.io/testing/) and paste me the results for `math.cos()` - don't forget to say the droid version :)

This set of values can determine OS. They're stable across ESR60, FF60-to-FF66, so test with any FF version you like. Definitely after some newer android version results and mac stuff (not iOS cuz thats the shitty webkit thing)

![mathcos](https://user-images.githubusercontent.com/16656956/53073888-37cd5400-34e1-11e9-8d9d-6937d793ded1.png)
^^ in case it's not clear, that just the last three digits that change

**Comment by Thorin-Oakenpants** [comment]
FYI: https://antoinevastel.com/browser fingerprinting/2019/02/19/canvas-fingerprint-on-the-web.html - aint read it yet . getting naked and drunk AF

**Comment by Thorin-Oakenpants** [comment]
hows everyone's BS entropy going?

Here's mine: FF66
![panopticlick-me](https://user-images.githubusercontent.com/16656956/57352565-4d076580-7155-11e9-83b7-f7fc192a9aec.png)

and here's TB8.0.9
![panopticlick-tb](https://user-images.githubusercontent.com/16656956/57352598-64465300-7155-11e9-9ac5-2ea2f69d4bdc.png)

Guess who's actually less FP'able?

**Comment by seanob86** [comment]
Here's mine: FF66 macOS

<img width="507" alt="Screen Shot 2019-05-08 at 18 39 30" src="https://user-images.githubusercontent.com/45988261/57365810-50124e00-7175-11e9-8222-7fdcf873b94f.png">

**Comment by Thorin-Oakenpants** [comment]
Excellent BS there @seanob86 ... I see you have your toolbar toggled on (height is fubared by a few pixels). Interesting font selection.

**Comment by seanob86** [comment]
OK thanks for the feedback regarding the toolbar. I've now disabled it and regarding the fonts I had in my overrides the following

user_pref("browser.display.use_document_fonts", 1);
user_pref("gfx.downloadable_fonts.woff2.enabled", true);

If I disable the above overrides and just use what's in the user.js file I now get the following

<img width="635" alt="Screen Shot 2019-05-09 at 08 47 05" src="https://user-images.githubusercontent.com/45988261/57415573-20525d00-71ec-11e9-8107-74c136526da2.png">

**Comment by Thorin-Oakenpants** [comment]
No need to change your toolbar being on/off. It's a bug: everyone else will be the same as you on Mac with a default toolbar density (and at least one bookmark)

**Comment by Thorin-Oakenpants** [comment]
Here's some more BS for you guys
- On the left is someone who uses the ghacks user.js and they posted it as an image on some forum
- On the right is me: I always block cookies: and the entropy for that is `6.29` and overall I am 1 in just over 4K with 11.9 entropy.

I noticed that this was the one thing that stood out as being higher. So I cleared all data, allowed a session cookie for panopticlick, and redid it = the pic below, on the right

In every **single** metric, my entropy **is lower** (except cookies where I am 0.01 higher, and language which is the same). And a few are earily close: i.e OPs image can't be very old (fourm post was a couple of days ago, and it looks like that's when he tested and posted it: also note the larger drop for the firefox header: his post got people who care about this shit (e.g firefox users), to test -> which drove down FF's entropy: self-fufilling prophecy shit all over again).

However, overall OP is `10.63`, and I am `14.7`. Look, overall entropy isn't adding up all the individual numbers (kinda, but not really). But linux (op) vs win32 (me) .. and 1000x900 (tor sizes) vs 1600x900 (pretty common screen res) - I mean come on. Tor users = 0.001%, linux users = 3% .. or whatever

What I'm really trying to say is that is really easy to poison/taint the data, and the whole thing is fucked ;) I can drive elephants thru almost every single figure on that site

![BS-entropy](https://user-images.githubusercontent.com/16656956/68522994-6e830e80-02aa-11ea-97a5-3d907cb072ee.png)

Have fun, kids .. until next time 🐘 🐘 🐘
